lib/TPP/GPU/GpuDataTransfer.cpp

```diff
@@ -0,0 +1,245 @@
+  if (auto viewOp = dyn_cast<mlir::ViewLikeOpInterface>(op))
+    return getDeviceTransferableBuffer(viewOp.getViewSource());
+
+  // Nothing to do for all the other cases.
+  // Assume it is a valid buffer, if the value comes from a device allocation,
+  // a function call, a function arguments, device allocation etc.
+  return nullptr;
+}
+
+// Matches the kernel operand and device buffer types such that data can be
+// transfered correctly from the original host buffer.
+// Returns failure if the two types cannot be connected.
+static LogicalResult matchTransferBuffers(RewriterBase &rewriter,
+                                          Value kernelOperand,
+                                          Value &gpuBuffer) {
+  auto loc = gpuBuffer.getLoc();
+
+  auto operandType = kernelOperand.getType().cast<MemRefType>();
+  auto gpuAllocType = gpuBuffer.getType().cast<MemRefType>();
+
+  // Kernel operands type already matches the device buffer.
+  // The host and device buffers can be used directly.
+  if (operandType == gpuAllocType)
+    return success();
+
+  // Examine view source type. More type conversions might be needed.
+  if (auto viewOp =
+          dyn_cast<mlir::ViewLikeOpInterface>(kernelOperand.getDefiningOp())) {
+    if (failed(matchTransferBuffers(rewriter, viewOp.getViewSource(),
+                                    gpuBuffer))) {
+      return failure();
+    }
+  }
+
+  // Cast device allocation to match the operand if possible.
+  if (memref::CastOp::areCastCompatible(gpuAllocType, operandType)) {
+    gpuBuffer =
+        rewriter.create<memref::CastOp>(loc, operandType, gpuBuffer).getDest();
+    return success();
+  }
+
+  // Take an equal subview of the device buffer if the operand is a subview.
+  // This ensures correct data access and eliminates need to change kernel
+  // argument types.
+  if (auto subview =
+          dyn_cast<memref::SubViewOp>(kernelOperand.getDefiningOp())) {
+    gpuBuffer = rewriter
+                    .create<memref::SubViewOp>(
+                        loc, operandType, gpuBuffer, subview.getOffsets(),
+                        subview.getSizes(), subview.getStrides(),
+                        subview.getStaticOffsets(), subview.getStaticSizes(),
+                        subview.getStaticStrides())
+                    .getResult();
+    return success();
+  }
+
+  // No way to connect the device buffer to the kernel operand.
+  return failure();
+}
+
+// Transfer host allocated data between the host and a device.
+// Optionally, copies data back to the host after device computations.
+// Returns device-allocated buffer with the moved data.
+static FailureOr<Value> transferMemref(RewriterBase &rewriter,
+                                       gpu::LaunchFuncOp launchFuncOp,
+                                       Value operand, Value hostBuffer,
+                                       bool copyDataBack = true) {
+  // A memref buffer is expected.
```

Comment:
I would assert here.

Reply:
Added assert for quicker feedback while debugging.